compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -421,6 +422,15 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         self.stack().last().map_or(self.tcx.span, |f| f.current_span())
     }
 
+    #[inline(always)]
+    /// Find the first stack frame that is within the current crate, if any, otherwise return the crate's HirId
+    pub fn best_lint_scope(&self) -> hir::HirId {
+        self.stack()
+            .iter()
+            .find_map(|frame| frame.body.source.def_id().as_local())
+            .map_or(CRATE_HIR_ID, |def_id| self.tcx.hir().local_def_id_to_hir_id(def_id))
+    }
```

Comment:
Miri has its own logic for this... so this is pretty const_eval specific I think, and should be in the const_eval folder, not the general interpreter.

Reply:
Also, are we sure this is the lint scope we want?
For regular errors we always use the 'root' scope I think, from where the const was declared. Why use a different one here?

(Though also with that root scope I wonder about associated consts where 'declared' and 'instantiated' are separate spans... we are only showing one of them in the errors currently, I think, and caching will make it hard to take into account the 'instantiated' span.)

Reply:
There is no root scope yet. I could add it, but it does fall fowl of associated/generic consts from other crates as you noted

Reply:
We have a root scope [here](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_const_eval/interpret/eval_context/struct.InterpCx.html#structfield.tcx).

Reply:
The reason I walk the stack is that even using a root const from another crate, via generics onr of your own consts or const fns may show up in the stack, and you have control over that

Reply:
Yeah I get that, but it seems a bit random -- slight changes in MIR building might lead to a completely different stacktrace after N terminators. I think I'd prefer if we always only considered the root, I am just not sure if we can even do that properly for generics from other crates. Having to allow this on a per-crate level is not great either so a fallback to the entire local crate is suboptimal...